Feature Event Scheduling app

izaz98 opened the pull request:
# Description:
This is an simple calender apps which supports adding, editing and deleting events. Moreover, it can also show the calender view of any month.

This app is design in following ways:
<h1>Design</h1>

There are 3 model classes.
<ul>
  <li>User</li>
  <li>Calender</li>
  <li>Event</li>
</ul>

The top level class in this heirarchy is User who has a calender and the calender has the events on it. The idea is prettry general and natural. In real life a user has a calender and he can add events by marker on his calender.

<h1>Flow</h1>

When we run the app a menu is displayed with different options. The user can perform different actions like adding, editing or deleting events by seleting the input from given menu.

izaz98 committed:
added Event class to store information related Events and added some basic code to check the valid arguments passed through command

izaz98 committed:
created helper.rb file, this will contain all the helper functions like show_month. Moreover, craeted show_month function in helper.rb and called it from the viewm case

izaz98 committed:
added code to create event and store it in events array.
override to_S function in Event class and added view_events function in helper to view all the events

izaz98 committed:
added check in viewa swtich statement for invalid arguments.
added a new statement where we are setting the month day as 1 when displaying month view

izaz98 committed:
moved from console commands to console I/O

izaz98 committed:
fixed liniting issues

izaz98 committed:
converted to MVC

izaz98 committed:
updated readme

arslanyaqoob7 commented on lib/controller/helper.rb:
```diff
@@ -0,0 +1,62 @@
+require_relative '../model/calender'
+require_relative '../model/event'
+
+def validate_date_time(time)
```
move to module

arslanyaqoob7 commented on lib/controller/helper.rb:
```diff
@@ -0,0 +1,62 @@
+  end
+end
+
+def add_event(user, title, description, time, index)
```
rename variable time -> date_time or date

arslanyaqoob7 commented on lib/model/calender.rb:
```diff
@@ -0,0 +1,44 @@
+    if @events.length <= 0
+      puts 'No events found'
+    else
+      events.each.with_index { |event, index| print index + 1, '- ', event.show_event, "\n" }
```
with_index -> with_index(1)

arslanyaqoob7 commented on lib/model/calender.rb:
```diff
@@ -0,0 +1,44 @@
+    puts month_view_header
+    print "\t" * (month_start_day - 1)
+    1.upto(num_of_days) do |num|
+      print num, (month_start_day % 7).zero? || num == num_of_days ? "\n" : "\t"
```
we can show no of events along with date in calendar view.

arslanyaqoob7 commented on lib/model/event.rb:
```diff
@@ -0,0 +1,20 @@
+class Event
+  attr_accessor :title, :description, :time, :date, :year, :day
+
+  def initialize(title, description, time, date, year, day, month)
```
rubocop -> use max 5 parameters in method definition.

arslanyaqoob7 commented on lib/view/easy_schedule.rb:
```diff
@@ -0,0 +1,70 @@
+    events_count = user.events_length
+
+    if events_count.positive?
+      index = validate_index_input(true, events_count)
```
user named variable.

arslanyaqoob7 commented on lib/view/easy_schedule.rb:
```diff
@@ -0,0 +1,70 @@
+    print 'Enter month (Jan 2021): '
+    month_to_print = gets.chop
+
+    month = Date.parse(month_to_print)
```
rename variable

arslanyaqoob7 commented on lib/model/calender.rb:
```diff
@@ -0,0 +1,44 @@
+  attr_accessor :cur_year, :cur_month, :events
+
+  def initialize(cur_month, cur_year, events)
+    @cur_month = cur_month
```
rename

izaz98 committed:
addressed comments: 1) created View class, 2) moved helper functions to module and included them in View class
3) used descriptive name for variables
4) followed some rubocop suggestions related function paramteres

arslanyaqoob7 commented on lib/controller/helper.rb:
```diff
@@ -0,0 +1,59 @@
+    puts 'Please enter the time in correct order \"3 Jan 2021 04:04 AM\"'
+  end
+
+  def add_event(user, title, description, event_date_time, index = nil)
```
method name should be add_or_update_event.

arslanyaqoob7 commented on lib/view/easy_schedule.rb:
```diff
@@ -0,0 +1,78 @@
+          index = validate_index_input(is_edit: true, events_count: events_count)
+
+          print "---Enter Event's new details---"
+          title, description, time = input_event_details
```
time validation are missing

arslanyaqoob7 commented on lib/view/easy_schedule.rb:
```diff
@@ -0,0 +1,78 @@
+      choice = gets.chop
+      case choice
+      when '1'
+        title, description, event_date_time = input_event_details
```
No validation on date input

arslanyaqoob7 commented on lib/model/calender.rb:
```diff
@@ -0,0 +1,58 @@
+    month_start_day = date.cwday
+    num_of_days = DateTime.new(date.year, date.month, -1).day
+    print month_view_header, "\n", "\t" * (month_start_day - 1)
+    1.upto(num_of_days) do |num|
```
num should be descriptive variable. i.e day

arslanyaqoob7 commented on lib/controller/helper.rb:
```diff
@@ -0,0 +1,59 @@
+  def validate_date_time(event_date_time)
+    DateTime.parse(event_date_time)
+  rescue ArgumentError
+    puts 'Please enter the time in correct order \"3 Jan 2021 04:04 AM\"'
```
Only showing an error message to the user, No validation

izaz98 committed:
added validations for string dates

izaz98 committed:
added index number validations
